crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
         Ok(())
     }
 
+    /// Look up the IBC denomination from a IbcToken.
+    pub fn query_ibc_denom(
+        &self,
+        token: impl AsRef<str>,
+        owner: Option<&Address>,
+    ) -> Result<String> {
+        let hash = match Address::decode(token.as_ref()) {
+            Ok(Address::Internal(InternalAddress::IbcToken(hash))) => {
+                hash.to_string()
+            }
+            _ => return Ok(token.as_ref().to_string()),
+        };
+
+        if let Some(owner) = owner {
+            let ibc_trace_key = ibc_trace_key(owner.to_string(), &hash);
+            if let Some(ibc_denom) = self.ctx.read_pre(&ibc_trace_key)? {
+                return Ok(ibc_denom);
+            }
+        }
+
+        // No owner is specified or the owner doesn't have the token
+        let ibc_denom_prefix = ibc_trace_key_prefix(None);
+        let ibc_denoms = self.ctx.iter_prefix(&ibc_denom_prefix)?;
+        for (key, ibc_denom, _gas) in ibc_denoms {
+            if let Some((_, token_hash)) =
+                is_ibc_trace_key(&Key::parse(key).into_storage_result()?)
+            {
+                if token_hash == hash {
+                    return String::try_from_slice(&ibc_denom[..])
+                        .into_storage_result()
+                        .map_err(Error::NativeVpError);
+                }
+            }
+        }
+
+        Ok(token.as_ref().to_string())
+    }
+
+    // Find the given IBC message in the changed keys and return the associated
+    // sequence number
+    fn search_ibc_transfer(
+        &self,
+        message: &IbcMsgTransfer,
+        keys_changed: &BTreeSet<Key>,
+    ) -> Result<Option<Sequence>> {
+        // Compute the packet commitment for this message
+        let packet_data_bytes = serde_json::to_vec(&message.packet_data)
+            .map_err(native_vp::Error::new)?;
+        let packet_commitment =
+            VpValidationContext::<'a, 'a, S, CA>::compute_packet_commitment(
+                &packet_data_bytes,
+                &message.timeout_height_on_b,
+                &message.timeout_timestamp_on_b,
+            );
+        // Try to find a key change with the same port, channel, and commitment
```

Comment:
Is there a reason we can't just construct the path directly and read from it (instead of iterating)? What data are we missing?

Reply:
We're missing the sequence number. Also, Yuji suggested a different approach:
> packet sequence can be obtained from the state with read_pre and the key next_sequence_*_key because the transaction has only one IBC message.

Reply:
Ah, I see.